compress/compress.go

```diff
@@ -0,0 +1,477 @@
+		bit := d.ReadBit()
+		xorLength = xorLength | (bit << i)
+		i--
+	}
+
+	j := uint64(xorLength)
+	var xor uint64
+	xor = 0
+	for uint32(j) != MaxUint32 {
+		bit := d.ReadBit()
+		xor = xor | (uint64(bit) << j)
+		j--
+	}
+
+	var rebuiltNumber uint64
+	rebuiltNumber = 0
+	previousBits := math.Float64bits(previous)
+	xor = xor << (64 - leadingZeros - xorLength)
+	rebuiltNumber = previousBits ^ xor
+
+	var buffer *bytes.Buffer
+	buffer = new(bytes.Buffer)
+	binary.Write(buffer, binary.BigEndian, rebuiltNumber)
+
+	b := buffer.Bytes()
+	readbuf := bytes.NewReader(b)
+	var result float64
+	err := binary.Read(readbuf, binary.BigEndian, &result)
+	if err != nil {
+		panic("WUT")
+	}
+
+	d.lengthOfWindow = uint64(xorLength)
+	d.leadingZeroWindowLength = uint64(leadingZeros)
+
+	return result
+}
+
+func (c *CompressionBuffer) Bytes() []byte {
+	if !c.finalized {
+		panic("Attempted to read bytes from an unfinalized compression buffer")
+	}
+	return c.buffer.Bytes()
+}
+
+func NewCompressionBuffer() CompressionBuffer {
+	return CompressionBuffer{
+		buffer:    new(bytes.Buffer),
+		position:  0,
+		current:   0,
+		firstPass: true,
+	}
+}
+
+func (c *CompressionBuffer) writeFirst(x float64) {
+	err := binary.Write(c.buffer, binary.BigEndian, x)
+	if err != nil {
+		panic("WUT")
+	}
+}
+
+func (c *CompressionBuffer) writeOne() {
+	c.current = c.current << uint64(1)
+	c.current = c.current | uint64(1)
+	c.position++
+	c.fixup()
+}
+
+func (c *CompressionBuffer) writeZero() {
+	c.current = c.current << 1
+	c.position++
+	c.fixup()
+}
+
+func (c *CompressionBuffer) fixup() {
+	if c.position == uint32(64) {
+		err := binary.Write(c.buffer, binary.BigEndian, c.current)
+		if err != nil {
+			panic("WUT")
+		}
+		c.position = 0
+		c.current = uint64(0)
+	}
+}
+
+func (c *CompressionBuffer) encodeMeaningfulXOR(x uint64) {
+	leadingZeros := leadingZeros64(x)
+	trailingZeros := trailingZeros64(x)
+	length := uint64(64) - uint64(leadingZeros) - uint64(trailingZeros)
+
+	c.writeOne()
+
+	//Case A: The meaningful bits of this block have the same # of leading
+	//zeros and field size as the previous compressed float.
+	if uint64(leadingZeros) == c.leadingZeroWindowLength && uint64(length) == c.lengthOfWindow && c.lengthOfWindow != 0 {
+		c.writeZero()
+		meaningfulRegion := x >> (64 - c.leadingZeroWindowLength - c.lengthOfWindow)
+		c.writeMeaningfulPartialXOR(c.lengthOfWindow, meaningfulRegion)
+	} else {
+		//Case B: This float has a different # of leading zeros and/or size
+		//than the previous one.
+		c.writeOne()
+		meaningfulRegion := (x >> trailingZeros)
+		c.writeMeaningfulXOR(uint64(leadingZeros), length, meaningfulRegion)
+
+		c.lengthOfWindow = length
+		c.leadingZeroWindowLength = uint64(leadingZeros)
+	}
+}
+
+func (c *CompressionBuffer) writeMeaningfulPartialXOR(length uint64, meaningful uint64) {
+	i := uint32(length)
+	for i != MaxUint32 {
+		bit := (meaningful >> i) & 1
+		if bit == 1 {
+			c.writeOne()
+		} else {
+			c.writeZero()
+		}
```

Comment:
That highlights an interesting choice of style. I find in implementing scientific/numerical algorithms that I favor more overt duplication of code because it lets me read the paper and implement it with a common language for larger-block pieces. For instance, in this case the writePartialXOR I can effectively read it in one go.

As an exercise, since you said "hopefully make it easier to understand", try checking out the branch and making the modifications and see if it helps with understanding. It's a fairly simple algorithm, but understanding it code-first is going to be tricky.